discovery/openstack/openstack.go

```diff
@@ -88,173 +70,22 @@ func NewDiscovery(conf *config.OpenstackSDConfig) (*Discovery, error) {
-	}
-	client, err := openstack.NewComputeV2(provider, gophercloud.EndpointOpts{
-		Region: d.region,
-	})
-
-	if err != nil {
-		return nil, fmt.Errorf("could not create OpenStack compute session: %s", err)
-	}
-
-	opts := servers.ListOpts{}
-	pager := servers.List(client, opts)
-
-	tg = &config.TargetGroup{
-		Source: fmt.Sprintf("OS_%s", d.region),
-	}
-
-	pagerFIP := floatingips.List(client)
-	floatingIPList := make(map[string][]string)
-
-	err = pagerFIP.EachPage(func(page pagination.Page) (bool, error) {
-		result, err := floatingips.ExtractFloatingIPs(page)
-		if err != nil {
-			log.Warn(err)
-		}
-		for _, ip := range result {
-			// Skip not associated ips
-			if ip.InstanceID != "" {
-				floatingIPList[ip.InstanceID] = append(floatingIPList[ip.InstanceID], ip.IP)
-			}
-		}
-		return true, nil
-	})
-
-	if err != nil {
-		return nil, fmt.Errorf("could not describe floating IPs: %s", err)
-	}
-
-	err = pager.EachPage(func(page pagination.Page) (bool, error) {
-		serverList, err := servers.ExtractServers(page)
-		if err != nil {
-			return false, fmt.Errorf("could not extract servers: %s", err)
-		}
-
-		for _, s := range serverList {
-			if len(s.Addresses) == 0 {
-				log.Debugf("Got no IP address for instance %s", s.ID)
-				continue
-			}
-			labels := model.LabelSet{
-				openstackLabelInstanceID: model.LabelValue(s.ID),
-			}
-
-			for _, address := range s.Addresses {
-				md, ok := address.([]interface{})
-				if !ok {
-					log.Warn("Invalid type for address, expected array")
-					continue
-				}
-
-				if len(md) == 0 {
-					log.Debugf("Got no IP address for instance %s", s.ID)
-					continue
-				}
-
-				md1, ok := md[0].(map[string]interface{})
-				if !ok {
-					log.Warn("Invalid type for address, expected dict")
-					continue
-				}
-
-				addr, ok := md1["addr"].(string)
-				if !ok {
-					log.Warn("Invalid type for address, expected string")
-					continue
-				}
-
-				labels[openstackLabelPrivateIP] = model.LabelValue(addr)
-
-				addr = net.JoinHostPort(addr, fmt.Sprintf("%d", d.port))
-
-				labels[model.AddressLabel] = model.LabelValue(addr)
-
-				// Only use first private IP
-				break
-			}
-
-			if val, ok := floatingIPList[s.ID]; ok {
-				if len(val) > 0 {
-					labels[openstackLabelPublicIP] = model.LabelValue(val[0])
-				}
-			}
-
-			labels[openstackLabelInstanceStatus] = model.LabelValue(s.Status)
-			labels[openstackLabelInstanceName] = model.LabelValue(s.Name)
-			id, ok := s.Flavor["id"].(string)
-			if !ok {
-				log.Warn("Invalid type for instance id, excepted string")
-				continue
-			}
-			labels[openstackLabelInstanceFlavor] = model.LabelValue(id)
-
-			for k, v := range s.Metadata {
-				name := strutil.SanitizeLabelName(k)
-				labels[openstackLabelTagPrefix+model.LabelName(name)] = model.LabelValue(v)
-			}
-
-			tg.Targets = append(tg.Targets, labels)
-		}
-		return true, nil
-	})
-
-	if err != nil {
-		return nil, fmt.Errorf("could not describe instances: %s", err)
+	switch d.role {
```

Comment:
Why is this case statement here in the Run method instead of the constructor? I don't see the need to keep `openstack.Discovery` and instantiate new instance/host discovery on every `Run` call.